[popover] event.preventDefault() should not cancel popover light dismiss

rwlbuis opened the pull request:
#### afc536e4bc276f974e802f28e5440b316ce2736f
<pre>
[popover] event.preventDefault() should not cancel popover light dismiss
<a href="https://bugs.webkit.org/show_bug.cgi?id=254384">https://bugs.webkit.org/show_bug.cgi?id=254384</a>

Reviewed by Tim Nguyen.

Move light dismiss logic to PointerCaptureController::dispatchEvent so the event can&apos;t
be cancelled, also see:
<a href="https://github.com/whatwg/dom/pull/1117#issuecomment-1311928234">https://github.com/whatwg/dom/pull/1117#issuecomment-1311928234</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::pointerEventWillBeDispatched):

Canonical link: <a href="https://commits.webkit.org/262283@main">https://commits.webkit.org/262283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f15dc8f83a3f9d69ede0ad87c8b2d8b34a2ed59d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1067 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1096 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1135 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1680 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/943 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1176 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1141 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1074 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/998 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1562 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1038 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/994 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/980 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/973 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2056 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1018 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/953 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1003 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/274 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1035 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -363,6 +364,11 @@ void PointerCaptureController::dispatchEvent(PointerEvent& event, EventTarget* t
+    if (event.type() == eventNames().pointerdownEvent || event.type() == eventNames().pointerupEvent) {
+        if (auto* element = dynamicDowncast<HTMLElement>(target))
+            element->document().handlePopoverLightDismiss(event, *element);
+    }
```
We should probably run this at the end of `pointerEventWillBeDispatched` per https://github.com/w3c/pointerevents/pull/460/files

nt1m replied:
@rwlbuis To be clear, this should run last in `PointerCaptureController::pointerEventWillBeDispatched`.

rwlbuis replied:
Sorry, I missed this comment earlier indeed, fixed.

nt1m commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -384,16 +385,24 @@ void PointerCaptureController::pointerEventWillBeDispatched(const PointerEvent&
+    if (!isPointerdown) {
+        handlePopoverLightDismiss();       
         return;
+    }
```
I think we can change the logic a bit to avoid repeating the call to handlePopoverLightDismiss E.g.

```
if (event.pointerType() != touchPointerEventType()) {
   if (RefPtr capturingData = m_activePointerIdsToCapturingData.get(pointerId))
      capturingData->pointerIsPressed = isPointerdown;
} else if (isPointerDown) {
  auto capturingData = ...;
  [...]
}
element->document().handlePopoverLightDismiss(event, *element);

nt1m commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -384,16 +385,24 @@ void PointerCaptureController::pointerEventWillBeDispatched(const PointerEvent&
         return;
 
+    auto handlePopoverLightDismiss = [&]() {
+        if (auto* element = dynamicDowncast<HTMLElement>(target))
```
I realize this check isn't correct, we should just pass any target directly into `handlePopoverLightDismiss`. Light dismiss should happen when you click outside on any node, not just HTML elements.

nt1m commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "EventHandler.h"
 #include "EventNames.h"
 #include "EventTarget.h"
+#include "HTMLElement.h"
```
This isn't needed anymore

nt1m commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -389,27 +390,25 @@ void PointerCaptureController::pointerEventWillBeDispatched(const PointerEvent&
-    auto capturingData = ensureCapturingDataForPointerEvent(event);
-    capturingData->pointerIsPressed = true;
-    setPointerCapture(downcast<Element>(target), pointerId);
+    auto* element = downcast<Element>(target);
```
`setPointerCapture` can probably make use of the `element` variable too

nt1m replied:
`auto element = downcast<Element>(*target);` is also slightly nicer, and `&element` if you need to refer to the pointer.

rwlbuis committed:
[popover] event.preventDefault() should not cancel popover light dismiss
https://bugs.webkit.org/show_bug.cgi?id=254384

Reviewed by Tim Nguyen.

Move light dismiss logic to PointerCaptureController::dispatchEvent so the event can't
be cancelled, also see:
https://github.com/whatwg/dom/pull/1117#issuecomment-1311928234

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::pointerEventWillBeDispatched):

Canonical link: https://commits.webkit.org/262283@main
